# Mental Health Question Columns Check

This notebook checks for missing values and anomalies in mental health related columns for Wave 2 and Wave 3 datasets.

**Targets:**
- **W2**: `v55_1` to `v55_14` in `C:\Users\user\Desktop\TIGPS_PLAN_DATA\TIGPS\20251229_new_progress\Data\2024data\TIGPS_W2_studentdata_ver3_encoded.csv`
- **W3**: `54-1` to `54-14` in `C:\Users\user\Desktop\TIGPS_PLAN_DATA\TIGPS\20251229_new_progress\Data\2025data\TIGPS_W3_student_studentdata_ver2_cleaned_cols.csv`

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
def check_columns(df, cols, dataset_name):
    print(f"--- Checking {dataset_name} ---")
    print(f"Total rows: {len(df)}")
    
    # Check if columns exist
    missing_cols = [c for c in cols if c not in df.columns]
    if missing_cols:
        print(f"WARNING: The following columns were not found in the dataframe: {missing_cols}")
        cols = [c for c in cols if c in df.columns]
    
    if not cols:
        print("No columns to check.")
        return

    # Missing values
    missing_counts = df[cols].isnull().sum()
    missing_pct = (missing_counts / len(df)) * 100
    
    print("\nMissing Values:")
    missing_df = pd.DataFrame({'Missing Count': missing_counts, 'Missing %': missing_pct})
    print(missing_df)
    
    print("\nValue Counts & Anomalies Check:")
    for col in cols:
        # Drop NaN for unique and checking anomalies
        valid_data = df[col].dropna()
        unique_vals = sorted(valid_data.unique())
        print(f"\nColumn: {col}")
        print(f"Unique values: {unique_vals}")
        # Check for values outside typical likert range
        print(df[col].value_counts().sort_index())

## Check W2 Data (`v55_1` - `v55_14`)

In [3]:
w2_path = r"C:\Users\user\Desktop\TIGPS_PLAN_DATA\TIGPS\20251229_new_progress\Data\2024data\TIGPS_W2_studentdata_ver3_encoded.csv"

if os.path.exists(w2_path):
    try:
        # Attempt to read with default engine, fall back if needed
        df_w2 = pd.read_csv(w2_path, low_memory=False)
        
        w2_cols = [f'v55_{i}' for i in range(1, 15)]
        check_columns(df_w2, w2_cols, "Wave 2 (W2)")
        
    except Exception as e:
        print(f"Error reading W2 file: {e}")
else:
    print(f"File not found: {w2_path}")

--- Checking Wave 2 (W2) ---
Total rows: 8892

Missing Values:
        Missing Count  Missing %
v55_1              89   1.000900
v55_2              89   1.000900
v55_3              89   1.000900
v55_4              90   1.012146
v55_5              89   1.000900
v55_6              89   1.000900
v55_7              89   1.000900
v55_8              89   1.000900
v55_9              89   1.000900
v55_10             89   1.000900
v55_11             89   1.000900
v55_12             89   1.000900
v55_13             89   1.000900
v55_14             89   1.000900

Value Counts & Anomalies Check:

Column: v55_1
Unique values: [np.float64(0.0), np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0)]
v55_1
0.0    6125
1.0    1907
2.0     405
3.0     160
4.0     206
Name: count, dtype: int64

Column: v55_2
Unique values: [np.float64(0.0), np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0)]
v55_2
0.0    4944
1.0    2099
2.0     887
3.0     365
4.0     508
Name: count, dtype:

## Check W3 Data (`54-1` - `54-14`)

In [4]:
w3_path = r"C:\Users\user\Desktop\TIGPS_PLAN_DATA\TIGPS\20251229_new_progress\Data\2025data\TIGPS_W3_student_studentdata_ver2_cleaned_cols.csv"

if os.path.exists(w3_path):
    try:
        df_w3 = pd.read_csv(w3_path, low_memory=False)
        
        w3_cols = [f'54-{i}' for i in range(1, 15)]
        check_columns(df_w3, w3_cols, "Wave 3 (W3)")
        
    except Exception as e:
        print(f"Error reading W3 file: {e}")
else:
    print(f"File not found: {w3_path}")

--- Checking Wave 3 (W3) ---
Total rows: 7714

Missing Values:
       Missing Count  Missing %
54-1               0        0.0
54-2               0        0.0
54-3               0        0.0
54-4               0        0.0
54-5               0        0.0
54-6               0        0.0
54-7               0        0.0
54-8               0        0.0
54-9               0        0.0
54-10              0        0.0
54-11              0        0.0
54-12              0        0.0
54-13              0        0.0
54-14              0        0.0

Value Counts & Anomalies Check:

Column: 54-1
Unique values: [np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]
54-1
1.0    4864
2.0    1817
3.0     556
4.0     245
5.0     232
Name: count, dtype: int64

Column: 54-2
Unique values: [np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]
54-2
1.0    4370
2.0    1830
3.0     777
4.0     357
5.0     380
Name: count, dtype: int64

Column: 54-

# Analysis Findings (2026-01-06)

### Wave 2 (W2) Performance
- **File**: `TIGPS_W2_studentdata_ver3_encoded.csv`
- **Columns**: `v55_1` to `v55_14`
- **Total Rows**: 8892
- **Missing Values**: Nearly consistent 89 missing values (~1.00%) for all columns.
- **Value Range**: **0.0 to 4.0** (5-point scale, 0-indexed).
- **Anomalies**: No numerical anomalies found outside the 0-4 range.

### Wave 3 (W3) Performance
- **File**: `TIGPS_W3_student_studentdata_ver2_cleaned_cols.csv`
- **Columns**: `54-1` to `54-14`
- **Total Rows**: 7714
- **Missing Values**: **0 missing values (0.00%)** for all 14 columns.
- **Value Range**: **1.0 to 5.0** (5-point scale, 1-indexed).
- **Anomalies**: No numerical anomalies found outside the 1-5 range.

### **Crucial Note on Data Consistency**
- **Scale Mismatch**: 
    - W2 uses a **0-4** scale.
    - W3 uses a **1-5** scale.
    - Before longitudinal analysis, these scales must be harmonized (e.g., mapping W3's 1-5 to 0-4 by subtracting 1, or vice versa).
